### PR TITLE
feat: Features.CustomStats

### PR DESCRIPTION
**Pull request by Misfiy**
## Description
**Describe the changes** 
Add Features.CustomStats

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?** (if this is a feature change)
Ability to set MaxStamina & MaxHumeShield

**Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

**Other information**:

<br />

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentations
<br />

## Submission checklist

- [ ] I have checked the project can be compiled
- [ ] I have tested my changes and it worked as expected

### Patches (if there are any changes related to Harmony patches)
- [ ] I have checked no IL patching errors in the console

### Other
- [X] Still requires more testing

### CODE REVIEW

**Review by VALERA771 — APPROVED** [review]

File: EXILED/Exiled.API/Features/CustomStats/CustomHumeShieldStat.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        /// Gets or sets the maximum amount of HumeShield the player can have.
+        /// </summary>
+        public float CustomMaxValue { get; set; } = -1;
+
+        private float ShieldRegeneration => TryGetHsModule(out HumeShieldModuleBase controller) ? controller.HsRegeneration * ShieldRegenerationMultiplier : 0;
+
+        /// <inheritdoc/>
+        public override void Update()
+        {
+            if (MaxValue == default)
```
Comment (Misfiy) [inline EXILED/Exiled.API/Features/CustomStats/CustomHumeShieldStat.cs:39]:
```suggestion
            if (MaxValue == -1)
```

File: EXILED/Exiled.API/Features/CustomStats/CustomStaminaStat.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    /// </summary>
+    public class CustomStaminaStat : StaminaStat
+    {
+        /// <inheritdoc/>
+        public override float MaxValue => CustomMaxValue == default ? base.MaxValue : CustomMaxValue;
+
+        /// <summary>
+        /// Gets or sets the maximum amount of stamina the player will have.
+        /// </summary>
+        public float CustomMaxValue { get; set; }
```
Comment (Misfiy) [inline EXILED/Exiled.API/Features/CustomStats/CustomStaminaStat.cs:25]:
```suggestion
        public float CustomMaxValue { get; set; } = -1;
```

File: EXILED/Exiled.API/Features/CustomStats/CustomStaminaStat.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    using PlayerStatsSystem;
+    using UnityEngine;
+
+    /// <summary>
+    /// A custom version of <see cref="StaminaStat"/> which allows the player's maximum amount of Stamina to be changed.
+    /// </summary>
+    public class CustomStaminaStat : StaminaStat
+    {
+        /// <inheritdoc/>
+        public override float MaxValue => CustomMaxValue == default ? base.MaxValue : CustomMaxValue;
```
Comment (Misfiy) [inline EXILED/Exiled.API/Features/CustomStats/CustomStaminaStat.cs:20]:
```suggestion
        public override float MaxValue => CustomMaxValue == -1 ? base.MaxValue : CustomMaxValue;
```

File: EXILED/Exiled.API/Features/CustomStats/CustomStaminaStat.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        /// <returns>The clamped num.</returns>
+        public float Clamp(float value) => CustomMaxValue == default ? Mathf.Clamp01(value) : Mathf.Clamp(value, 0, MaxValue);
+
+        /// <summary>
+        /// Overiding NW Method to sync Player percentage of Stamina.
+        /// </summary>
+        /// <param name="writer">The writer.</param>
+        public override void WriteValue(NetworkWriter writer)
+        {
+            if (CustomMaxValue == default)
```
Comment (Misfiy) [inline EXILED/Exiled.API/Features/CustomStats/CustomStaminaStat.cs:40]:
```suggestion
            if (CustomMaxValue == -1)
```

File: EXILED/Exiled.API/Features/CustomStats/CustomStaminaStat.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+            writer.WriteByte(ToByte(CurValue / CustomMaxValue));
+        }
+
+        /// <summary>
+        /// Overriding NW Method to sync Player percentage of Stamina.
+        /// </summary>
+        /// <param name="reader">The reader.</param>
+        /// <returns>the float value sync to server.</returns>
+        public override float ReadValue(NetworkReader reader) => CustomMaxValue == default ? base.ReadValue(reader) : CurValue * CustomMaxValue;
```
Comment (Misfiy) [inline EXILED/Exiled.API/Features/CustomStats/CustomStaminaStat.cs:54]:
```suggestion
        public override float ReadValue(NetworkReader reader) => CustomMaxValue == -1 ? base.ReadValue(reader) : CurValue * CustomMaxValue;
```

**Comment by Misfiy** [comment]
Thats things i should work make it better (to allow 0 stamina & 0 shield)

File: EXILED/Exiled.API/Features/CustomStats/CustomHumeShieldStat.cs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+                new SyncedStatMessages.StatMessage()
+                {
+                    Stat = this,
+                    SyncedValue = CurValue,
+                }.SendToHubsConditionally(CanReceive);
+                _lastSent = CurValue;
+                _valueDirty = false;
+            }
+
+            if (ShieldRegeneration == 0)
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/CustomStats/CustomHumeShieldStat.cs:59]: Cool be interesting to override this value

File: EXILED/Exiled.Events/Patches/Fixes/FixStaminaBar.cs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    using Exiled.API.Features.CustomStats;
+    using HarmonyLib;
+    using Mirror;
+    using PlayerRoles.FirstPersonControl;
+    using UnityEngine;
+
+    /// <summary>
+    /// Fix for <see cref="CustomStaminaStat"/>.
+    /// </summary>
+    [HarmonyPatch(typeof(FpcStateProcessor), nameof(FpcStateProcessor.UpdateMovementState))]
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Fixes/FixStaminaBar.cs:19]: I am pretty sure there a way to not having to patch this event

**Review by louis1706 — CHANGES_REQUESTED** [review]

File: EXILED/Exiled.API/Features/CustomStats/CustomHumeShieldStat.cs (hunk 6 above)
Comment (Misfiy) [reply EXILED/Exiled.API/Features/CustomStats/CustomHumeShieldStat.cs:59]: I made a multiplier for it

File: EXILED/Exiled.API/Features/CustomStats/CustomStaminaStat.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        /// Gets or sets the maximum amount of stamina the player will have.
+        /// </summary>
+        public float CustomMaxValue { get; set; } = -1;
+
+        /// <summary>
+        /// Clamps a float to fit the current stamina bar.
+        /// </summary>
+        /// <param name="value">The value to clamp.</param>
+        /// <returns>The clamped num.</returns>
+        public float Clamp(float value) => CustomMaxValue == default ? Mathf.Clamp01(value) : Mathf.Clamp(value, 0, MaxValue);
```
Comment (Misfiy) [inline EXILED/Exiled.API/Features/CustomStats/CustomStaminaStat.cs:32]:
```suggestion
        public float Clamp(float value) => CustomMaxValue == -1 ? Mathf.Clamp01(value) : Mathf.Clamp(value, 0, MaxValue);
```

File: EXILED/Exiled.Events/Patches/Fixes/FixStaminaBar.cs (hunk 7 above)
Comment (Misfiy) [reply EXILED/Exiled.Events/Patches/Fixes/FixStaminaBar.cs:19]: Should prob be transpiler but idk

**Comment by Misfiy** [comment]
This needs testing, I made it work previously but this is a redo of it so not 100% itll work

File: EXILED/Exiled.Events/Patches/Generic/FixStaminaStat.cs [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+                __result = state;
+                return false;
+            }
+
+            stat.CurValue = stat.Clamp(stat.CurValue + (__instance.ServerRegenRate * Time.deltaTime));
+            __result = state;
+            return false;
+        }
+    }
+}
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/FixStaminaStat.cs:73]: I will review later this on why it's needed

**Review by louis1706 — COMMENTED** [review]
I will review later on why this is needed

File: EXILED/Exiled.Events/Patches/Generic/FixStaminaStat.cs (hunk 9 above)
Comment (Misfiy) [reply EXILED/Exiled.Events/Patches/Generic/FixStaminaStat.cs:73]: we need to make it not clamp from 0 to 1 (a transpiler would probably work better here)

File: EXILED/Exiled.API/Features/CustomStats/CustomHumeShieldStat.cs [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        private float ShieldRegeneration => TryGetHsModule(out HumeShieldModuleBase controller) ? controller.HsRegeneration * ShieldRegenerationMultiplier : 0;
+
+        /// <inheritdoc/>
+        public override void Update()
+        {
+            if (MaxValue == -1)
+            {
+                base.Update();
+                return;
+            }
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/CustomStats/CustomHumeShieldStat.cs:43]: && ShirldRegenerationMultiplier is 1

**Review by louis1706 — CHANGES_REQUESTED** [review]

File: EXILED/Exiled.Events/Patches/Generic/FixStaminaStat.cs (hunk 9 above)
Comment (VALERA771) [reply EXILED/Exiled.Events/Patches/Generic/FixStaminaStat.cs:73]: Idk about what is better now. We are changing game logic pretty much. We can use transpiler only to call method, if needed, that will set/return values to set. Changing all of these stuff to transpiler looks too complex to update

File: EXILED/Exiled.API/Features/CustomStats/CustomHumeShieldStat.cs [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        /// Gets or sets the maximum amount of HumeShield the player can have.
+        /// </summary>
+        public float CustomMaxValue { get; set; } = -1;
+
+        private float ShieldRegeneration => TryGetHsModule(out HumeShieldModuleBase controller) ? controller.HsRegeneration * ShieldRegenerationMultiplier : 0;
+
+        /// <inheritdoc/>
+        public override void Update()
+        {
+            if (MaxValue == -1)
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/CustomStats/CustomHumeShieldStat.cs:39]:
```suggestion
            if (MaxValue == -1 && ShieldRegenerationMultiplier is 1)
```

**Review by louis1706 — APPROVED** [review]

**Pull request closed without merging**